doc/dev/schema.md

```diff
@@ -0,0 +1,193 @@
+This information is stored on storages in some tables and Tarantool
+spaces: ``_ddl_sharding_key`` and ``_ddl_sharding_func``. ``crud``
+module uses `_ddl_sharding_key` and `_ddl_sharding_func` spaces to fetch
+sharding schema: thus, you don't obliged to use ``ddl`` module and can
+setup only ``_ddl_*`` spaces manually, if you want. ``crud`` module uses
+plain Lua tables to store sharding info on routers and storages.
+
+### When schema is used
+
+Sharding schema is used to compute ``bucket_id``. Thus, it is used each
+time we need to execute a non-map-reduce request (``insert_*``,
+``replace_*``, ``update_*``, ``delete_*``,  ``upsert_*``, ``get_*``
+requests, most ``select`` and ``count`` requests, some ``pairs``
+requests), ``bucket_id`` is not specified explicitly and there is no
+``force_map_call`` option. If there is no sharding schema specified, we
+use defaults: sharding key is primary key, sharding function is
+``vshard.router.bucket_id_strcrc32``.
+
+### How schema is reloaded
+
+Storage sharding schema (internal ``crud`` Lua tables) is updated on
+initialization (first time when info is requested) and each time someone
+changes ``_ddl_sharding_key`` or ``_ddl_sharding_func`` data -- we use
+``on_replace`` triggers.
+
+Routers fetch sharding schema if cache wasn't initialized yet or each
+time reload was requested. Reload could be requested with ``crud``
+itself (see below) or by user (with ``require('crud.common.sharding_key').update_cache()``
+or ``require('crud.common.sharding_func').update_cache()`` handles).
+The handles was deprecated after introducing automatic reload.
+
+It is worth noting that reload/initialization affects both sharding
+key and function caches at the same time, but fetch itself is triggered
+even if one cache requested by some function is not initialized yet.
+(I.e. if function cache is ok and key cache is not, they still both be
+reloaded.). If there are several requests to reload info, the only one
+reload fiber is started and every request waits for its completion.
+
+### When schema is reloaded and operation is retried
+
+Each request that uses sharding info passes sharding hashes from
+a router with a request. If hashes mismatch with storage ones, we return
+a specific error. The request retries if it receives a hash mismatch
+error from the storage. If it didn't help after ``N`` retries (now ``N``
+is ``1``), the error is passed to the user. For ``*_many`` methods,
+reload and retry happens only if all tuples had failed with hash
+mismatch; otherwise, errors are passed to the user.
+
+### Alternative approaches
+
+There were different implementations of working with hash: we tried
+to compute it instead of storing pre-computed values, but pre-computed
+approach with triggers was better in terms of performance. It was also
+an option to ping storage before sending a request and verify sharding
+info relevance before sending the request with separate call, but it was
+discarded.
```

Comment:
Why?

Reply:
I mean, I understand why, but the alternatives are listed to explain why an approach that may look simpler/better was not choosen. If it is so obvious to don't mention, it is not an alternative and should not be written here. If it is an alternative, it worth to say why it is worse than choosen approach.

Reply:
Added a reason